allure-generator/src/main/java/io/qameta/allure/allure2/Allure2ResultsReader.java

```diff
@@ -99,45 +96,45 @@ public Allure2ResultsReader(AttachmentsStorage storage) {
                         dest.setTestStage(testStage);
                     }
 
-                    dest.getBeforeStages().addAll(getParentBefores(result.getParentIds(), groups));
-                    dest.getBeforeStages().addAll(convert(result.getBefores(), this::convert));
-                    dest.getAfterStages().addAll(convert(result.getAfters(), this::convert));
-                    dest.getAfterStages().addAll(getParentAfters(result.getParentIds(), groups));
+                    List<TestResultContainer> parents = findAllParents(groups, result.getId(), new HashSet<>());
+                    dest.getBeforeStages().addAll(getStages(parents, this::getBefore));
+                    dest.getAfterStages().addAll(getStages(parents, this::getAfter));
                     return dest;
                 }).collect(Collectors.toList());
     }
 
-    private List<StageResult> getParentBefores(List<String> parents, Map<String, TestGroupResult> groups) {
-        return getFromParents(parents, (result, stageResults) -> {
-            stageResults.addAll(convert(result.getBefores(), this::convert));
-            return stageResults;
-        }, groups, new HashSet<>());
+    private List<StageResult> getStages(List<TestResultContainer> parents,
+                                        Function<TestResultContainer, Stream<StageResult>> getter) {
+        return parents.stream().flatMap(getter).collect(Collectors.toList());
     }
 
-    private List<StageResult> getParentAfters(List<String> parents, Map<String, TestGroupResult> groups) {
-        return getFromParents(parents, (result, stageResults) -> {
-            List<StageResult> current = convert(result.getAfters(), this::convert);
-            current.addAll(stageResults);
-            return current;
-        }, groups, new HashSet<>());
+    private Stream<StageResult> getBefore(TestResultContainer container) {
+        Comparator<StageResult> comparator = Comparator.comparingLong(a -> a.getTime().getStart());
+        return convert(container.getBefores(), this::convert).stream()
+                .sorted(comparator);
     }
 
-    private <T> List<T> getFromParents(List<String> parents, BiFunction<TestGroupResult, List<T>, List<T>> getter,
-                                       Map<String, TestGroupResult> groups, Set<String> seen) {
-        return parents.stream()
-                .flatMap(parentId -> getFromParents(parentId, getter, groups, seen).stream())
-                .collect(Collectors.toList());
+    private Stream<StageResult> getAfter(TestResultContainer container) {
```

Comment:
seems like for `befores` / `afters` there could be created a generic method with accepts a `list` and `comparator`